### ISSUE TITLE
Transaction status notification using RPC

### ISSUE THREAD

**Issue by lyulka** [parallelchain-io/parallelchain-protocol#13]
# Transaction event notifications and the Transaction Courier

### Definitions

- **Submitted Transaction**: a transaction that has been submitted to a Fullnode using the `submit_transaction` RPC.
- **Committed Transaction**: a transaction that has been included in a committed block.
- **Pending Transaction**: a submitted transaction that has not become committed but is deemed to have a *high probability* of being committed in the future within a reasonable timeframe and without re-submission.
- **Lost Transaction**: a submitted transaction that has not become committed and is deemed to have only a *low probability* of being committed in the future within a reasonable timeframe and without re-submission.

### Purpose

A transaction successfully submitted using `submit_transaction` may never progress to being included a block that is eventually committed for any of a large set of possible reasons.

A client that have been waiting for some time for a submitted transaction to be committed needs a facility to distinguish between two scenarios:
1. The transaction is lost--in which case the client should consider resubmitting it.
2. The transaction is *not* lost ("pending"), and the network is just very busy and higher priority transactions are being included in blocks first--in which case the client should just sit and wait.

This issue describes a design for such a facility. 

### Design principles

The probability that a transaction will become included in a block in the future without re-submission cannot be determined exactly without a global knowledge of complete internal states of all replicas and clients in the system. Further, it is hard to define exactly what counts as a "small probability" in deciding whether a transaction is lost or not.

In an effort to keep things *simple*, this design does not attempt to get such a global knowledge, nor does is it based on a precise definition of what counts as a "small probability". Instead, this design works on the principle of best effort, and what "small probability" means in this context is "small enough that I should tell the client to consider re-submitting".

### Architecture

```mermaid
graph BT;

%% Processes
fullnode["<b>Fullnodes (many)</b><br/> <ul><li>/subscribe_to_transaction_events</li></ul>"];

courier[<b>Transaction Courier</b><br/> <ul><li>/submit_transaction</li><li>/transaction_status</li><li>/dispatched_transactions</li></ul>];

explorer_be["<b>Explorer (back-end)</b> <ul><li>/submit_transaction</li>"];
explorer_fe["<b>Explorer (front-end)</b>"];

%% Relations
courier--WebSocket---fullnode;
explorer_be-->courier;
explorer_fe-->explorer_be;
```

This design comprises two sets of additions:
1. In **Fullnode RPC**: 
    - A new `subscribe_to_transaction_events` RPC that creates a WebSocket that reports events about changes to the "position" of a transaction as specified by a filter.
2. **Transaction Courier**: a new web service that Explorer (back-end) submits transactions to. It "dispatches" transactions to a Fullnode, and then keeps track of their statuses by listening on transaction status subscriptions. It has three RPCs:
    - `submit_transaction`: causes courier to dispatch the transaction to a Fullnode's RPC.
    - `dispatched_transactions`: returns a list of recently-dispatched transactions along with their current ["status"](#status-function).
    - `transaction_status`: returns a specific transaction's status.`

#### Request

```rust
struct SubscribeToTransactionEventsRequest {
    filter: Option<Filter>,
}

enum Filter {
    TransactionHash(HashSet<CryptoHash>),
    Signer(HashSet<PublicAddress>),
}
```

#### Response

None. [Notification WebSocket](#notification-websocket) established.

### Notification WebSocket

The Notification WebSocket supports three types of messages:
- **Notification**: containing `(TransactionHash, Event)`.
- [**Ping** and **Pong**](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#pings_and_pongs_the_heartbeat_of_websockets): each side may send a ping to the other side, expecting a pong to test connection and/or process liveness. 

### Events

An Event is an enum whose variants indicate that "something has happened" to the Transaction:

```rust
enum TransactionEvent {
    Received,
    InsertedToMempool,
    RejectedFromMempool(RejectedFromMempoolReason),
    RemovedFromMempool(RemovedFromMempoolReason),
    PoppedToExecutor,
    IncludedInBlock,
    RejectedFromBlock(RejectedFromBlockReason),
    BlockPruned,
    BlockCommitted(TransactionHash),
}

enum RejectedFromMempoolReason {
    NonceLTCommitted,
    BaseFeePerGasTooLow,
    MempoolIsFull,
    Other,
}

enum RemovedFromMempoolReason {
    NonceLTCommitted,
    BaseFeePerGasTooLow,
    MempoolIsFull,
    Other,
}

enum RejectedFromBlockReason {
    NonceLTCommitted,
    NonceNESpeculative,
    BaseFeePerGasTooLow,
    BalanceLTInclusionCost,
    Other,
}
```

The set of possible variants of `TransactionEvent` is deliberately large, and some reasons are deliberately vague, this is so that different Fullnode implementations and configurations have the flexibility to implement different transaction pipelines or to perform stricter or looser "checks" on Transactions. 

For example, what "too low" in `BaseFeePerGasTooLow` can be dynamic: a Fullnode may maintain a statistic of the lowest n-percentile base fee per gas in blocks of the past 1 month, and reject transactions whose base fee per gas are below this.

### Dispatched Transactions Table

The dispatched transactions table is a set of three-tuples that keep track of the events relating to the transactions that the transaction courier has dispatched recently. This table can be be stored in memory, or in a persistent database:

|Field|Type|
|---|---|
|Transaction Hash|`CryptoHash`|
|Received|`SystemTime`|
|Latest Events|`HashMap<FullnodeID, CourierEvent>`|

#### Courier Events

The "Latest Events" field of the dispatched transactions table hold `CourierEvent`s. This is an enum with all of the variants of [`TransactionEvent`](#events), but with one additional variant `PresumedLostInCrash` that is added to the map if a specific Fullnode is presumed to have crashed (e.g., the Courier timed out while waiting for them to send a Pong in reply to a Ping).

```rust
enum CourierEvent {
    /* TransactionEvent variants omitted */
    PresumedLostInCrash,
}
```

#### Clearing

Old (according to Received time) entries in the dispatched transactions table must be removed. How long an entry remains in the table before it is removed should be a configurable and/or dynamic parameter (e.g., if the network is busy and it is taking a long time for a transaction to be included in a block, entries can be made to remain for longer).

### Status function

The Status enum indicates whether a transaction is [pending or lost](#definitions):

```rust
enum Status {
    Pending,
    Lost,
}
```

The courier predicts the status of a dispatched transaction, on the basis of [best effort](#design-principles), by considering the Latest Events received for that particular transaction as stored in the dispatched transaction.

The mapping between latest events and a status can be implemented as a function:

```rust
fn status(latest_events: Vec<CourierEvent>) -> Status;
```

The next subsection gives some guidance on how to implement this function.

#### Lost-ness of events

We categorize variants of Event into "auto-lost events" and "threshold-lost events" depending on how strongly receiving one instance of each variant indicates that a transaction is lost:
- **Auto-lost event**: the event strongly indicates that a transaction has been lost. An example is the `RejectedFromMempool` event with a `NonceLTCommitted` reason. This indicates that the specific transaction's nonce is lower than the signer account's nonce in the committed world state. Since the nonce in the committed world state is strictly increasing, this transaction can never be included in a future block.
- **Threshold-lost event** the event only weakly indicates that the transaction has been lost. An example is the `RejectedFromMempool` event with a `MempoolIsFull` reason. A transaction being dropped from one Fullnode's mempool does not hurt its probability of being included in a future block in a reasonable timeframe that much unless this event is received from *many* Fullnodes.

Threshold-lost Events events are named such because having received a specific number ("threshold") of such events may be taken as indication that the transaction has been lost.

We provide a tentative categorization of event variants below:

##### Auto-lost events

- `RejectedFromMempool` with reasons:
    - `NonceLTCommitted`
    - `BaseFeePerGasTooLow`
    - `Other`
- `RemovedFromMempool` with reasons:
    - `NonceLTCommitted`
    - `BaseFeePerGasTooLow`
    - `Other`
- `RejectedFromBlock` with reasons:
    - `NonceLTCommitted`
    - `NonceNESpeculative`

##### Threshold-lost events

- `RejectedFromMempool` with reasons:
    - `MempoolIsFull`
- `RemovedFromMempool` with reasons:
    - `MempoolIsFull`
- `BlockPruned`.

#### `submit_transaction`

Dispatches the transaction included in the request to a Fullnode's `submit_transaction` RPC.

##### Request and Response

Identical to Fullnode RPC's `submit_transaction`.

#### `dispatched_transactions`

List dispatched transactions and their [statuses](#status-function). 

##### Request

```rust
struct DispatchedTransactionsRequest {
    skip: u32,
    limit: u32,
}
```

##### Response

```rust
struct DispatchedTransactionsResponse {
    transactions: Vec<(Transaction, Status)>,
}
```

#### `transaction_status`

List a particular transaction's status.

##### Request

```rust
struct TransactionStatusRequest {
    transaction_hash: CryptoHash,
}
```

##### Response

```rust
struct TransactionStatusResponse {
    transaction_hash: CryptoHash,
    status: Option<Status>,
}
```

If a matching entry is not found in the dispatched transaction table, status is `None`.

## Open design questions

1. The Courier may receive a transaction from its `submit_transaction` RPC that has already been committed. How to handle it?
2. Should the Courier put in its table transactions that it didn't personally dispatch.
3. Exact delineation of events into auto-lost and threshold-lost.
4. Exact value for threshold.
5. When to start removing entries in the Courier table.
6. What kind of HotStuff-rs API can we use to find out committed and pruned blocks.
7. What kind of interface should there be between Executor and Mempool and WS server.
8. How to handle the load of lots of persistent WS connections?

**Comment by TLiu2084** [comment]
Proposed Design 1

## Data Structure
``` rust
// Mapping of transaction and list of clients that want to receive tx status notice.
type Clients = Arc<Mutex<HashMap<TxHash, Vec<UnboundedSender<Event>>>>;

// RPC request struct defined in protocol
// a `notify` field is added
struct SubmitTransactionRequest {
    transaction: Transaction,
    notify: boolean,
}

// Server-Sent Event with transaction status
struct Event {
    tx_hash: SHA256,
    status: TransactionStatusNotification,
}

enum TransactionStatusNotification {
    // Txn is inserted to mempool
    Received,
    // Txn cannot be inserted to mempool
    Rejected(RejectReason),
    // Txn is inserted to mempool but being dropped later on
    Dropped(DropReason),
    // Txn is included in committed block
    Committed(BlockHash),
}

// RejectReason specifies the possible reasons that transaction cannot be added to mempool.
enum RejectReason {
    GasLimitTooLow,
    GasLimitGreaterThanBlockGasLimit,
    BaseFeeTooLow,
    TxSizeTooLarge,
    WrongSignature,
}

// DropReason specifies the possible reasons that transaction cannot be added to committed block.
enum DropReason {
    NonceTooLow,
    NonceTooHigh,
    SignerBalanceTooLow,
    Staled,
}
```
1. When recevie SubmitTransactionRequest from RPC., reply with `Server-Sent Events (SSE)` with KeepAlive stream to maintain the liveness of event source connection. 
```rust
// [POST] /submit_transaction
fn submit_transaction_route {
    let clients: Clients = Arc::new(Mutex::new(Hashmap::new()));
    let with_clients = warp::any().map(move || clients.clone());
    
    warp::path!("submit_transaction")
        .and(warp::post())
        .and(warp::body::content_length_limit(10 * 1024 * 1024))
        .and(warp::body::bytes())
        .and(with_clients)
        .map(|input_bytes, clients: Clients| {
            // reply using server-sent events
            let stream = txn_received(clients);
            warp::sse::reply(warp::sse::keep_alive().stream(stream))
      });
}
```
2. Create a unbounded channel for each incoming transaction to receive Event from source (Executor / Mempool). If the txn.notify is enabled, push the channel sender to client list.
```rust
fn txn_received(
    input_bytes: Bytes,
    clients: Clients,
) -> impl Stream<Item = Result<Event, warp::Error>> + Send + 'static {
    // Use an unbounded channel to handle buffering and flushing of messages to the event source...
    let (tx, rx) = mpsc::unbounded_channel();
    let rx = UnboundedReceiverStream::new(rx);

    // Try to deserialize input_bytes into Transaction
    let txn = Transaction::deserialize(input_bytes);
    
    if deserialize success {
    	if txn.notify == true {
	    // Save the sender in our list of connected users.
	    clients.lock().unwrap().insert(tx_hash, tx);
        }
    
        // send transaction to mempool
        match submit_transaction(txn){
            Ok() => tx.send(Event with `Received` status),
            Err(e) => tx.send(Event with `Rejected` status)
        }
    }
    else {
    	tx.send(Event with `Rejected` status)
    }

    // Convert event messages into warp `Server-Sent Events (SSE)` and return resulting stream.
    rx.map(|event| {
        Ok(Event::default()
            .event("txn_notice")
            .json_data(event.to_json())
            .unwrap())
    })
}
```
3. Setup a thread to keep listen the message from Executor and Mempool. Use mpsc channel to send messages between threads Executor/Mempool and push Event to connected user channels.
```rust
// Setup mpsc channel used by Executor/Mempool
let (status_tx, status_rx) = mpsc::channel();

fn send_notification(status_rx: Receiver<(tx_hash, TransactionStatusNotification)>, clients: Clients){
    tokio::runtime::Handle::current().spawn(async move {
        loop {
            if let Some((tx_hash, status)) = status_rx.recv().await {
               // Send event if there are clients subscribed notification of that transaction status.
               // Remove the client list of corresponding tx_hash
		if Some(Vec<UnboundedSender<Event>>) = clients.lock().unwrap().remove(tx_hash){
		   for client_tx in Vec<UnboundedSender<Event>>{
		   	client_tx.send(Event with corresonding transaction status)
		   }
		}
            }
        }
    })
}
```
## Questions (tbc)
- What if a transaction gets included in a block, but then the branch that block is in is pruned off? (i.e., a conflicting branch becomes committed).

- If a transaction is dropped because a particular replica’s mempool is full, or the transaction is staled in that particular mempool, should the transaction courier mark it as “dropped”?

**Comment by TLiu2084** [comment]
### Questions
> 1. The Courier may receive a transaction from its submit_transaction RPC that has already been committed. How to handle it?

Courier should forward all transactions that it received to Fullnode, and wait for fullnode RPC notification. Here are two reasons:
 - Extra communication from Fullnode is needed to determine committed transaction. As the entries in dispatched transactions table would be removed after a period of time, courier may not have enough information to determine the transaction status.
- Fullnode should reject the committed transaction and send out `RejectedFromMempool` with reason `NonceLTCommitted`.

> 2. Should the  put in its table transactions that it didn't personally dispatch.

Yes, from users' perspective, they want a way to check their transaction status even they're not submitting transaction by explorer api and courier. Courier would setup a websocket connection without filter to receive notifications from those Fullnodes it subscribed.

> 3. Exact delineation of events into auto-lost and threshold-lost.

TBC

> 4. Exact value for threshold.

TBC

> 5. When to start removing entries in the Courier table.

TBC

> 6. What kind of HotStuff-rs API can we use to find out committed and pruned blocks.

Refer to new version HotStuff-rs design. 

> 7. What kind of interface should there be between Executor and Mempool and WS server.

Refer to the [Design details in Fullnode RPC](https://github.com/parallelchain-io/parallelchain-protocol/issues/13#issuecomment-1689287517), we use unbounded mspc channel to communicate between Executor and Mempool and WS server.
It spawns a thread to receive `TransactionEvent` from Executor and Mempool and forward to the clients that has been subscribed to that transaction status. 

> 8. How to handle the load of lots of persistent WS connections?

TBC

**Comment by lyulka** [comment]
A couple brief comments @TLiu2084:

- My initial idea for the API that HotStuff-rs can provide is some kind of global event bus. In special points in the HotStuff-rs source code, I will send a message through a channel to this bus. Users can attach closures to this bus to act as event handlers. This generalizes the logging mechanism that HotStuff-rs already has. The logger can be an event handler which attaches to this bus.
- The categorization of events into “auto-lost” and “threshold-lost” is only a suggestion. Maybe this categorization is too broad and simple, and each should event maps into a status in an individual way.

**Comment by TLiu2084** [comment]
> A couple brief comments @TLiu2084:
> 
> * My initial idea for the API that HotStuff-rs can provide is some kind of global event bus. At special points in the HotStuff-rs source code, I will send a message through a channel to this bus. Users can attach closures to this bus to act as event handlers. This generalizes the logging mechanism that HotStuff-rs already has. The logger can be an event handler which attaches to this bus.

@lyulka ok, that means we need another event handler thread in Fullnode listen to this channel and  relay the transaction status through websocket.

> * The categorization of events into “auto-lost” and “threshold-lost” is only a suggestion. Maybe this categorization is too broad and simple, and each should event maps into a status in an individual way.

In current design, there are only 2-3 drop reasons may cause false alarm that users' transaction is actually not dropped globally. I think it is fair enough to group the dropped events into two categories only.

**Comment by lyulka** [comment]
Merged into ParallelChain protocol v0.5 changelog #8.

**Comment by lyulka** [comment]
Will track updates to this enhancement in #8.
